pymatgen/entries/compatibility.py

```diff
@@ -407,9 +411,127 @@ def __str__(self):
         return "{} {} Correction".format(self.name, self.compat_type)
 
 
-class Compatibility(MSONable):
+class Compatibility(MSONable, metaclass=abc.ABCMeta):
     """
-    The Compatibility class combines a list of corrections to be applied to
+    Abstract Compatibility class, not intended for direct use.
+    Compatibility classes are used to correct the energies of an entry or a set
+    of entries. All Compatibility classes must implement .get_adjustments method.
+    """
+    @abc.abstractmethod
+    def get_adjustments(self, entry: ComputedEntry):
+        """
+        Get the energy adjustments for a ComputedEntry.
+
+        This method must generate a list of EnergyAdjustment objects
+        of the appropriate type (constant, composition-based, or temperature-based)
+        to be applied to the ComputedEntry, and must raise a CompatibilityError
+        if the entry is not compatible.
+
+        Args:
+            entry: A ComputedEntry object.
+
+        Returns:
+            [EnergyAdjustment]: A list of EnergyAdjustment to be applied to the
+                Entry.
+
+        Raises:
+            CompatibilityError if the entry is not compatible
+        """
+        return
+
+    def process_entries(self, entries: Union[ComputedEntry, list], clean: bool = False):
+        """
+        Process a sequence of entries with the chosen Compatibility scheme.
+
+        Args:
+            entries: ComputedEntry or [ComputedEntry]
+            clean: bool, whether to remove any previously-applied energy adjustments.
+                If True, all EnergyAdjustment are removed prior to processing the Entry.
+                Default is False.
+
+        Returns:
+            A list of adjusted entries.  Entries in the original list which
+            are not compatible are excluded.
+        """
+        # convert input arg to a list if not already
+        if isinstance(entries, ComputedEntry):
+            entries = [entries]
+
+        processed_entry_list = []
+
+        if clean:
+            for e in entries:
+                # remove any previously documented corrections from the entry
+                e.energy_adjustments = []
+
+        # apply the corrections
+        for entry in entries:
+            try:
+                adjustments = self.get_adjustments(entry)
+                for ea in adjustments:
```

Comment:
I think you're doing too much checking here. There is some expectation that the implenter knows what they're doing.

Reply:
The checking is more for the user than the implementer. To avoid unexpected behavior, I think it's really important that we enforce `EnergyAdjustment` only being added once by any given class. I should be able to run `Compatibility.process_entries` 10 times over the same set of entries and get the same result as if I only ran it once. I know running 10 times would not be intended use, but it could easily happen inadvertently with different pieces of code talking to each other (e.g., user gets processed entries from MPRester and then tries to reprocess themselves)

Reply:
You're using one tool to accomplish a complex thing. Protecting users beyond sensible defaults is not easy. If the user wants to break your logic they will. The harder they have to fight your checks, the worse the break will be. It's best to have safe sensible defaults that will encourage the appropriate behavior, and if the user decides to do something stupid, that is on them.

In this case, you have another way to enforce good practices via `clean`. Set `clean` to be default `True` and document this in the docstrings. If the user or implementer decides to set `clean` to `False`, then they are on their own.

Reply:
I agree we can only put in so many guardrails for users doing 'stupid' things. But 'stupid' is different from 'not obvious and unintentionally wrong'. I want to provide enough of a safety net to prevent at least a few 'not obvious' unintended behaviors that have personally tripped me up. What do you think is the harm in too much checking?

Enforcing this via `clean=True` is going to conflict with your comment to preserve `ManualEnergyAdjustment`. To maintain backward compatibility, if I instantiate a `ComputedEntry` with `correction = XXX`, that creates a `ManualEnergyAdjustment`. If we preserve those when `clean=True`, then we're going to have problems with legacy entries.

Reply:
That sounds like a problem with `from_dict`. It should treat the `dict` correction as a ConstantCorrection. Only those coming in via `setter` should create a `ManualEnergyAdjustment`

Reply:
> That sounds like a problem with `from_dict`. It should treat the `dict` correction as a ConstantCorrection. Only those coming in via `setter` should create a `ManualEnergyAdjustment`

I thought one of the fundamental tenets of `as_dict` / `from_dict` was that you should get back exactly the same object when you call ` from_dict` ? Wouldn't this break that principle?

Reply:
The objects should be equivalent, but in this case we're updating an old representation to a new one, so its allowed. Backward compatibility is important when possible.

Reply:
Yes, definitely agree that we need to preserve backward compatibility here. I feel like a safer way around all this though is to have `clean=True` strip out `ManualEnergyAdjustment` along with any other. That's more straightforward to understand, and it avoids all this ambiguity about what "manual" really means. It seems to me that anytime as_dict / from_dict results in ANY change to the object, there is a huge potential for confusion and bugs.

I'll note that the current behavior of `MaterialsProjectCompatibility` is effectively the same as `clean=True`. So users should be accustomed to manually re-applying manual corrections if they reprocess entries.